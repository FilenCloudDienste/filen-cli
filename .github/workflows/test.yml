name: Run tests

on:
  push:
  pull_request:
    types: [opened, synchronize, reopened]

jobs:
  test:
    runs-on: ubuntu-latest
    env:
      FILEN_CLI_TESTING_EMAIL: ${{ secrets.V2_TEST_ACC_EMAIL }}
      FILEN_CLI_TESTING_PASSWORD: ${{ secrets.V2_TEST_ACC_PASS }}
    steps:
      - name: Checkout
        uses: actions/checkout@v3
      - name: Install bun
        uses: oven-sh/setup-bun@v2
      - run: bun install
      - run: bun run test