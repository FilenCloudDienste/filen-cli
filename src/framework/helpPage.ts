import { formatTable } from "./util"
import { Extra, Feature, FeatureGroup, OptionArgument, PositionalArgument } from "./features"
import { App } from "./app"

export function printHelp<X extends Extra>(app: App<X>, selectedName: string, isInteractiveMode: boolean) {
	const selectedFeature = (() => {
		if (selectedName.length === 0) return app.features.featureGroup
		const featureGroup = app.features.getFeatureGroup(selectedName)
		if (featureGroup) return featureGroup
		return app.features.findFeature(selectedName)?.feature
	})()
	if (!selectedFeature) {
		app.outErr(`Unknown command or help topic: ${selectedName}`)
		return
	}

	const builder = new HelpTextBuilder()
	const printFeatureHelp = (feature: Feature<X> | FeatureGroup<X>) => {
		// recursively print FeatureGroup
		if (Object.hasOwn(feature, "features")) { // is FeatureGroup
			const featureGroup = feature as FeatureGroup<X>
			const isSelected = featureGroup.name !== undefined && featureGroup.name === (selectedFeature as FeatureGroup<X>).name
			if (!isSelected && featureGroup.visibility === "hide") return
			builder.appendNewline()
			if (!isSelected && featureGroup.visibility === "collapse") {
				builder.appendNewline()
				builder.appendText((() => {
						if (featureGroup.title) return featureGroup.title
						if (featureGroup.name) return featureGroup.name
						if (featureGroup.features.length > 0 && Object.prototype.hasOwnProperty.call(featureGroup.features[0], "cmd")) {
							return (featureGroup.features[0] as Feature<X>).cmd[0]!
						}
						return "***"
					})())
				const referenceName = featureGroup.name !== undefined ? featureGroup.name
					: featureGroup.features.length === 0 ? undefined
					: Object.prototype.hasOwnProperty.call(featureGroup.features[0], "cmd") ? (featureGroup.features[0] as Feature<X>).cmd[0]!
					: (featureGroup.features[0] as FeatureGroup<X>).name
				if (referenceName) builder.appendText(` (expand via \`filen help ${referenceName}\`)`)
				builder.appendNewline()
				return
			}
			if (featureGroup.title) {
				builder.appendText(featureGroup.title)
			}
			if (featureGroup.description) {
				if (featureGroup.features.length === 0) builder.appendNewline()
				builder.appendText(featureGroup.description)
			}
			if (featureGroup.longDescription) {
				builder.withIncreasedIndentation(() => {
					builder.appendNewline()
					builder.appendText(featureGroup.longDescription!)
					builder.appendNewline()
				})
			}
			builder.appendNewline()
			if (featureGroup.features.length > 0) {
				builder.appendNewline()
				const printFeatures = () => featureGroup.features.forEach(feature => printFeatureHelp(feature))
				if (featureGroup.title) {
					builder.withIncreasedIndentation(() => printFeatures())
				} else {
					printFeatures()
				}
			}
			return
		}

		// print Feature command signature and description
		feature = feature as Feature<X>
		const isOptional = (arg: PositionalArgument<X> | OptionArgument<X>) => arg.kind === "option" && !(arg as OptionArgument<X>).isRequired
		builder.appendText("> " + [feature.cmd[0],...feature.arguments.map(arg => `${isOptional(arg) ? "[" : "<"}${arg.name}${arg.kind === "catch-all" ? "..." : ""}${isOptional(arg) ? "]" : ">"}`)].join(" "))
		builder.withIncreasedIndentation(() => {
			if (feature.description) {
				builder.appendText(feature.description!)
			}
			if (feature.longDescription) {
				builder.appendNewline()
				builder.appendText(feature.longDescription!)
				builder.appendNewline()
			}
			const formatArguments = feature.arguments.filter(arg => arg.description !== undefined).map(arg => ({ name: `${arg.name}${arg.kind === "catch-all" ? "..." : ""}`, description: arg.description, optional: isOptional(arg) }))
			if (formatArguments.length > 0) {
				builder.appendText(formatTable(formatArguments.map(arg => [`${arg.optional ? "[" : "<"}${arg.name}${arg.optional ? "]" : ">"}`, arg.description ?? ""])))
			}
		})
		
		builder.appendNewline()
	}
	if (!isInteractiveMode) {
		builder.appendText(app.info.name + " " + app.info.version)
	}
	printFeatureHelp(selectedFeature)
	builder.print(app)
}

class HelpTextBuilder<X extends Extra> {
	private lines: { text: string, indentation: number }[] = []
	private indentation = 0

	appendText(text: string) {
		this.lines.push({ text, indentation: this.indentation })
	}

	appendNewline() {
		if (this.lines[this.lines.length-1]?.text !== "") {
			this.lines.push({ text: "", indentation: this.indentation })
		}
	}

	withIncreasedIndentation(callback: () => void) {
		this.indentation++
		callback()
		this.indentation--
	}

	print(app: App<X>) {
		this.lines.forEach(line => {
			app.out(line.text, { indentation: line.indentation })
		})
	}
}