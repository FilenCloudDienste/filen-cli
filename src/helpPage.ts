export const helpPage: string = `
Usage: filen [options...] <cmd...>

Options:
${formatTable([
	["--help, -h", "display usage information"],
	["--verbose, -v", "display additional information"],
	["--email <email>", "specify credentials"],
	["--password <password>", ""],
	["--two-factor-code <code>, -c <code>", "(optional)"],
	["--quiet, -q", "hide things like progress bars"],
	["--root <path>, -r <path>", "execute a stateless command from a different working directory"],
	["--json", "format output as JSON"],
	["--no-autocomplete", "disable autocompletion (for performance or bandwidth reasons)"],
])}

Commands:
${formatTable([
	["ls [dir]", "list items inside a directory"],
	["cat <file>", "print content of a text file"],
	["mkdir <dir>", "create a directory"],
	["rm <path>", "delete a file or directory"],
	["download <cloud path> <local path>", "download a file or directory from the cloud into a local destination"],
	["upload <local file> <cloud path>", "upload a local file into the cloud at a specified path"],
	["stat <path>", "display information about a file or directory"],
	["statfs", "display information about your Filen cloud drive"],
	["mv <path from> <path to>", "move a file to a path (parent directory or file)"],
	["cp <path from> <path to>", "copy a file to a path (parent directory or file)"],
	["write <file> <content...>", "write text to a file"],
	["open <file>", "open a file locally in the associated application"],
	["edit <file>", "edit a file locally in the associated application (save and close to re-upload)"],
])}

Interactive mode:
Invoke the Filen CLI without any specified commands to enter interactive mode. 
${formatTable([
	["cd <path>", "navigate to a different path"],
	["ls", "list items inside current directory"],
	["exit, ^C", "exit interactive mode"]
])}

Read the full documentation at: https://github.com/FilenCloudDienste/filen-cli#readme
`

/**
 * Formats a two-dimensional array as a table.
 * @param table An array of rows
 * @param spacing Amount of whitespaces between columns
 */
function formatTable(table: string[][], spacing: number = 3): string {
	let columns = 0
	table.forEach(row => {
		if (row.length > columns) columns = row.length
	})

	const columnWidths: number[] = []
	for (let i = 0; i < columns; i++) {
		columnWidths.push(0)
		table.forEach(row => {
			const cell = row[i]
			if (cell.length > columnWidths[i]) columnWidths[i] = cell.length
		})
	}

	const lines: string[] = []
	table.forEach(row => {
		const line: string[] = []
		for (let column = 0; column < row.length; column++) {
			line.push(row[column] + " ".repeat(columnWidths[column] - row[column].length))
		}
		lines.push(line.join(" ".repeat(spacing)))
	})
	return lines.join("\n")
}