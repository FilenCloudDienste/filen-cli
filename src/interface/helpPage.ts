import dedent from "dedent"
import { version } from "../buildInfo"

/**
 * Provides general and specific help pages.
 */
export class HelpPage {
	/**
	 * @param topic "general" or a specific topic
	 * @return the help page for a topic
	 */
	getHelpPage(topic: string): string | undefined {
		if (topic === "general") return this.generalHelpPage
		if (topic === "fs") return this.fsHelpPage
		if (topic === "sync") return this.syncHelpPage
		if (topic === "webdav") return this.webdavHelpPage
		if (topic === "s3") return this.s3HelpPage
		return undefined
	}

	/**
	 * @return the help page shown when using the `help` command in interactive mode
	 */
	getInteractiveModeHelpPage() {
		return this.interactiveModeHelpPage
	}

	private readonly versionUrlSegment = version === "0.0.0" ? "" : `/tree/${version}`

	private readonly commands = [
		["ls [dir]", "list items inside a directory"],
		["cat <file>", "print content of a text file"],
		["mkdir <dir>", "create a directory"],
		["rm <path>", "delete a file or directory"],
		["download <cloud path> <local path>", "download a file or directory from the cloud into a local destination"],
		["upload <local file> <cloud path>", "upload a local file into the cloud at a specified path"],
		["stat <path>", "display information about a file or directory"],
		["statfs", "display information about your Filen cloud drive"],
		["whoami", "print the current user"],
		["mv <path from> <path to>", "move a file to a path (parent directory or file)"],
		["cp <path from> <path to>", "copy a file to a path (parent directory or file)"],
		["write <file> <content...>", "write text to a file"],
		["open <file>", "open a file locally in the associated application"],
		["edit <file>", "edit a file locally in the associated application (save and close to re-upload)"],
	]
	private readonly interactiveModeCommands = [
		["cd <path>", "navigate to a different path"],
		["ls", "list items inside current directory"],
		["exit, ^C", "exit interactive mode"]
	]

	private readonly interactiveModeHelpPage: string = dedent`
		Commands:
		${formatTable([...this.commands, ...this.interactiveModeCommands])}}
		`

	private readonly generalHelpPage: string = dedent`
		Usage: filen [options...]
		
		Options:
		${formatTable([
			["--help, -h", "display usage information"],
			["--verbose, -v", "display additional information"],
			["--email <email>", "specify credentials"],
			["--password <password>", ""],
			["--two-factor-code <code>, -c <code>", "(optional)"],
		])}
		
		View the topic pages via \`filen -h <topic>\` for more information:
		${formatTable([
			["fs", "Access your Filen Drive"],
			["sync", "Syncing locations with the cloud"],
			["webdav", "WebDAV mirror server with single user or proxy mode"],
			["s3", "S3 mirror server"],
		])}
		
		Read the full documentation at: https://github.com/FilenCloudDienste/filen-cli${this.versionUrlSegment}#readme
		`

	private readonly fsHelpPage: string = dedent`
		Usage: filen [options...] <cmd...>
		
		Additional options:
		${formatTable([
			["--quiet, -q", "hide things like progress bars"],
			["--root <path>, -r <path>", "execute a stateless command from a different working directory"],
			["--json", "format output as JSON"],
			["--no-autocomplete", "disable autocompletion (for performance or bandwidth reasons)"],
		])}
		
		Commands:
		${formatTable(this.commands)}
		
		Interactive mode:
		Invoke the Filen CLI with no command specified to enter interactive mode. 
		${formatTable(this.interactiveModeCommands)}
		
		Read the full documentation at: https://github.com/FilenCloudDienste/filen-cli${this.versionUrlSegment}#access-your-filen-drive
		`

	private readonly syncHelpPage: string = dedent`
		Usage: filen sync [sync pairs...] [--continuous]
		
		Examples:
		${formatTable([
			["filen sync", "read sync pairs from $APP_DATA/filen_cli/syncPairs.json of type {local: string, remote: string, syncMode: string, alias?: string}[]"],
			["filen sync <file>", "read sync pairs from custom JSON file"],
			["filen sync mypair myotherpair", "use aliases as defined in syncPairs.json"],
			["filen sync /local/path:/cloud/path", "sync a local path with a cloud path in two-way sync"],
			["filen sync /local1::/cloud1 /local2:twoWay:/cloud2", "other ways to specify two-way sync"],
			["filen sync /local1:localToCloud:/cloud1 /local2:ltc:/cloud2", "localToCloud sync (other sync modes: cloudBackup, cloudToLocal, cloudBackup)"]
	    ])}
	    
		Set the --continuous flag to keep syncing (instead of only syncing once).
	    
		Read the full documentation at: https://github.com/FilenCloudDienste/filen-cli${this.versionUrlSegment}#syncing 
		`

	private readonly webdavHelpPage: string = dedent`
		Usage:
		  Single user: filen webdav --w-user <...> --w-password <...> [options...]
		  Proxy mode:  filen webdav-proxy [options...]
		
		Additional options:
		${formatTable([
			["--w-https", "use HTTPS instead of HTTP (using a self-signed certificate)"],
			["--w-hostname", "which hostname the server should be started on (default is 0.0.0.0)"],
			["--w-port", "which port the server should be started on (default is 80 or 443)"],
			["--w-auth-scheme", "the authentication scheme the server should use, \"basic\" or \"digest\" (default is basic)"],
		])}
		
		Read the full documentation at: https://github.com/FilenCloudDienste/filen-cli${this.versionUrlSegment}#webdav-server
		`

	private readonly s3HelpPage: string = dedent`
		Usage: filen s3 --s3-access-key-id <...> --s3-secret-access-key <...> [options...]
		
		Additional options:
		${formatTable([
			["--s3-https", "use HTTPS instead of HTTP (using a self-signed certificate)"],
			["--s3-hostname", "which hostname the server should be started on (default is 0.0.0.0)"],
			["--s3-port", "which port the server should be started on (default is 80 or 443)"],
		])}
		
		Read the full documentation at: https://github.com/FilenCloudDienste/filen-cli${this.versionUrlSegment}#s3-server
		`
}

/**
 * Formats a two-dimensional array as a table.
 * @param table An array of rows
 * @param spacing Amount of whitespaces between columns
 */
function formatTable(table: string[][], spacing: number = 3): string {
	let columns = 0
	table.forEach(row => {
		if (row.length > columns) columns = row.length
	})

	const columnWidths: number[] = []
	for (let i = 0; i < columns; i++) {
		columnWidths.push(0)
		table.forEach(row => {
			const cell = row[i]
			if (cell !== undefined && cell.length > columnWidths[i]!) columnWidths[i] = cell.length
		})
	}

	const lines: string[] = []
	table.forEach(row => {
		const line: string[] = []
		for (let column = 0; column < row.length; column++) {
			line.push(row[column] + " ".repeat(columnWidths[column]! - row[column]!.length))
		}
		lines.push(line.join(" ".repeat(spacing)))
	})
	return lines.join("\n")
}