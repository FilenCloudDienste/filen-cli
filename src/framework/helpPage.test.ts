import { describe, expect, it } from "bun:test"
import { printHelp } from "./helpPage"
import { mockFrameworkApp } from "../test/tests"

describe("help page", () => {

    const _app = () => mockFrameworkApp(f => [
        f.helpText({ name: "general", text: "This is general help." }),
        f.helpText({ name: "something", text: "This is something help." }),
        { features: [
            f.feature({
                cmd: ["my-feature"],
                description: "This is my feature.",
                longDescription: "This is a longer description of my feature.",
                args: {
                    myArg: f.arg({ name: "myArg", description: "This is my argument." }),
                    myOptionalArg: f.optionalArg({ name: "myOptionalArg", description: "This is my optional argument." }),
                    myDefaultValueArg: f.defaultValue("hello there", f.optionalArg({ name: "myDefaultValueArg", description: "This is my argument with a default value." })),
                    myCatchAllArg: f.catchAllArg({ name: "myCatchAllArg", description: "This is my catch-all argument." }),
                    myOptionArg: f.option({ name: "--myOptionArg", description: "This is my option argument." }),
                    myFlagArg: f.flag({ name: "--myFlagArg", alias: "-f", description: "This is my flag argument." }),
                    myRequiredOptionArg: f.required(f.option({ name: "--myRequiredOptionArg", description: "This is my required option argument." })),
                },
                invoke: async () => {},
            }),
        ], visibility: "collapse" },
        f.helpText({ title: "My Title", name: "my-title", text: "This is a collapsed help text description.", visibility: "collapse" }),
    ])
    
    it("should print general help", () => {
        const { app, output } = _app()
        printHelp(app, "", false)
        expect(output()).toMatchSnapshot()
    })

    it("should also print general help by specified name", () => {
        const { app, output } = _app()
        printHelp(app, "general", false)
        expect(output()).toMatchSnapshot()
    })

    it("should print any help by specified name", () => {
        const { app, output } = _app()
        printHelp(app, "something", false)
        expect(output()).toMatchSnapshot()
    })

    describe("collapsed sections", () => {
        it("should print in full when requested", () => {
            const { app, output } = _app()
            printHelp(app, "my-title", true)
            expect(output()).toMatchSnapshot()
        })
    })

    it("should print feature correctly", () => {
        const { app, output } = _app()
        printHelp(app, "my-feature", false)
        expect(output()).toMatchSnapshot()
    })

})