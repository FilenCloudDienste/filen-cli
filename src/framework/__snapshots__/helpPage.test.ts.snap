// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`help page collapsed sections should print in full when requested 1`] = `
"
My Title

This is a collapsed help text description.

"
`;

exports[`help page should print general help 1`] = `
"MyName 0.0.0

This is general help.

This is something help.

my-feature
 (expand via \`filen help my-feature\`)

My Title
 (expand via \`filen help my-title\`)

"
`;

exports[`help page should also print general help by specified name 1`] = `
"MyName 0.0.0

This is general help.

"
`;

exports[`help page should print any help by specified name 1`] = `
"MyName 0.0.0

This is something help.

"
`;

exports[`help page should print feature correctly 1`] = `
"MyName 0.0.0
> my-feature <myArg> [myOptionalArg] [myDefaultValueArg] <myCatchAllArg...> [--myOptionArg <...>] [--myFlagArg] <--myRequiredOptionArg <...>>
    This is my feature.
    
    This is a longer description of my feature.
    
    <myArg>                         This is my argument.                                            
    [myOptionalArg]                 This is my optional argument.                                   
    [myDefaultValueArg]             This is my argument with a default value. (default: hello there)
    <myCatchAllArg...>              This is my catch-all argument.                                  
    [--myOptionArg <...>]           This is my option argument.                                     
    [--myFlagArg]                   This is my flag argument.                                       
    <--myRequiredOptionArg <...>>   This is my required option argument.                            

"
`;
